pkg/kubelet/kuberuntime/kuberuntime_logs.go

```diff
@@ -0,0 +1,332 @@
+	}
+	if apiOpts.SinceTime != nil && apiOpts.SinceTime.After(opts.since) {
+		opts.since = apiOpts.SinceTime.Time
+	}
+	return opts
+}
+
+// ReadLogs read the container log and redirect into stdout and stderr.
+func ReadLogs(path string, apiOpts *api.PodLogOptions, stdout, stderr io.Writer) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return fmt.Errorf("failed to open log file %q: %v", path, err)
+	}
+	defer f.Close()
+
+	// Search start point based on tail line.
+	opts := newLogOptions(apiOpts, time.Now())
+	start, err := tail(f, opts.tail)
+	if err != nil {
+		return fmt.Errorf("failed to tail %d lines of log file %q: %v", opts.tail, path, err)
+	}
+	if _, err := f.Seek(start, os.SEEK_SET); err != nil {
+		return fmt.Errorf("failed to seek %d in log file: %v", start, path)
+	}
+
+	// Start parsing the logs.
+	r := bufio.NewReader(f)
+	// Do not create watcher here because it is not needed if `Follow` is false.
+	var w *fsnotify.Watcher
+	var parse parseFunc
+	msg := &logMessage{}
+	for {
+		l, err := r.ReadBytes(eol[0])
+		if err != nil {
+			if err != io.EOF { // This is an real error
+				return fmt.Errorf("failed to read log: %v", err)
+			}
+			if !opts.follow {
+				// Return directly when reading to the end if not follow.
+				if len(l) > 0 {
+					glog.Warningf("Incomplete line: %q", l)
+				}
+				return nil
+			}
+			// Reset seek so that if this is an incomplete line,
+			// it will be read again.
+			if _, err := f.Seek(-int64(len(l)), os.SEEK_CUR); err != nil {
+				return fmt.Errorf("failed to reset seek: %v", err)
+			}
+			if w == nil {
+				// Intialize the watcher if it has not been initialized yet.
+				if w, err = fsnotify.NewWatcher(); err != nil {
+					return fmt.Errorf("failed to create fsnotify watcher: %v", err)
+				}
+				defer w.Close()
+				if err := w.Add(f.Name()); err != nil {
+					return fmt.Errorf("failed to watch file %q: %v", f.Name(), err)
+				}
+			}
+			// Wait until the next log change.
+			if err := waitLogs(w); err != nil {
+				return fmt.Errorf("failed to wait logs: %v", err)
+			}
+			continue
+		}
+		if parse == nil {
+			// Intialize the log parsing function.
+			parse, err = getParseFunc(l)
+			if err != nil {
+				return fmt.Errorf("failed to get parse function: %v", err)
+			}
+		}
+		// Parse the log line.
+		msg.reset()
+		if err := parse(l, msg); err != nil {
+			glog.Errorf("Failed with err %v when parsing log: %q", err, l)
+			continue
+		}
+		// Write the log line into the stream.
+		if err := writeLogs(msg, stdout, stderr, opts); err != nil {
+			glog.Errorf("Failed with err %v when writing log: %+v", err, msg)
+		}
+	}
+}
+
+// parseFunc is a function parsing one log line to the internal log type.
+// Notice that the caller must make sure logMessage is not nil.
+type parseFunc func([]byte, *logMessage) error
+
+var parseFuncs []parseFunc = []parseFunc{
+	parseCRILog,        // CRI log format parse function
+	parseDockerJSONLog, // Docker JSON log format parse function
+}
```

Comment:
nit: Should doc those supported log formats (togather with delimiter and eol) in CRI interface.